lms/djangoapps/program_enrollments/api/writing.py

```diff
@@ -193,30 +197,59 @@ def write_program_course_enrollments(
     if not (create or update):
         raise ValueError("At least one of (create, update) must be True")
     requests_by_key, duplicated_keys = _organize_requests_by_external_key(enrollment_requests)
-    external_keys = set(requests_by_key)
+    processable_external_keys = set(requests_by_key)
+
+    results = {}
+    results.update({
+        key: ProgramCourseOpStatuses.DUPLICATED for key in duplicated_keys
+    })
+
+    if not processable_external_keys:
+        return results
+
     program_enrollments = fetch_program_enrollments(
         program_uuid=program_uuid,
-        external_user_keys=external_keys,
+        external_user_keys=processable_external_keys,
     ).prefetch_related('program_course_enrollments')
     program_enrollments_by_key = {
         enrollment.external_user_key: enrollment for enrollment in program_enrollments
     }
 
-    # Fetch existing program-course enrollments.
-    existing_course_enrollments = fetch_program_course_enrollments(
-        program_uuid, course_key, program_enrollments=program_enrollments,
+    # Fetch enrollments regardless of anchored Program Enrollments
+    existing_course_enrollments = fetch_program_course_enrollments_by_students(
+        external_user_keys=processable_external_keys,
+        course_keys=[course_key],
+    ).select_related('program_enrollment')
+
+    conflicting_user_key_and_status = _get_conflicting_active_course_enrollments(
+        requests_by_key,
+        existing_course_enrollments,
+        program_uuid,
+        course_key
+    )
+
+    # Remove the conflicted items from the requests dictionary,
+    # so we can continue our processing below
+    for conflicting_user_key in conflicting_user_key_and_status:
```

Comment:
Thanks for this doing this mutation here. I know it's more verbose, but I really think it'll be able to reason about in the future.

Reply:
Alex and your feedback make sense. It's easier to follow to do mutation here anyways.